### ISSUE TITLE
Season pack search not properly matching quality to particular season

### ISSUE THREAD

**Issue by bobbysteel** [SiCKRAGETV/sickrage-issues#2838]
Looking at the logs, it seems the season pack code doesn't want to properly map to the appropriate season. I see two issues - multiple (valid) season packs in a result set are fed into pickBestResult when they should be broken out by season and each season individually then fed into pickBestResult (line 499 in search.py)
Secondly, there seems to be a bug in line 521 where the SQL query does not properly match the exact season of the bestSeasonResult. This seems to originate in like 511 where the season is meant to be parsed out in the (x in episodes) bit.
I'd love to check in fixes but I'm not terribly familiar with the data structures here unfortunately like "episodes".

Supporting logs:

2015-09-17 11:37:08 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Using cached parse result for: Peep.Show.S07.720p.WEB-DL.AAC2.0.h.264-DOC
2015-09-17 11:37:08 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Found result Peep.Show.S07.720p.WEB-DL.AAC2.0.h.264-DOC at https://api.nzb.su/getnzb/1a544e7f198f44031bdfa7aaea6a8c15.nzb&i=9477&r=**********
2015-09-17 11:37:08 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Separating full season result to check for later
2015-09-17 11:37:08 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Using cached parse result for: Peep.Show.S07.720p.WEB-DL.AAC2.0.H.264-DOC
2015-09-17 11:37:08 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Found result Peep.Show.S07.720p.WEB-DL.AAC2.0.H.264-DOC at https://api.nzb.su/getnzb/36b1158f5fa93de5b7ad62ff0fa1593b.nzb&i=9477&r=**********
2015-09-17 11:37:08 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Separating full season result to check for later
2015-09-17 11:37:08 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Parsed Peep.Show.S06.720p.WEB-DL.AAC2.0.H.264-DOC into Peep Show - S06 [GROUP: DOC] [ABD: False] [ANIME: False] [whichReg: ['season_only']]
2015-09-17 11:37:08 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Found result Peep.Show.S06.720p.WEB-DL.AAC2.0.H.264-DOC at https://api.nzb.su/getnzb/ce8b6b65462a5f6565dd2c07d3bb4f95.nzb&i=9477&r=**********
2015-09-17 11:37:08 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Separating full season result to check for later
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Transaction with 48 queries executed
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Picking the best result out of [u'Peep.Show.S07.720p.WEB-DL.AAC2.0.h.264-DOC', u'Peep.Show.S07.720p.WEB-DL.AAC2.0.H.264-DOC', u'Peep.Show.S06.720p.WEB-DL.AAC2.0.H.264-DOC', u'Peep.Show.S04.SDTV.Xvid']
2015-09-17 11:37:09 INFO     SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Quality of Peep.Show.S07.720p.WEB-DL.AAC2.0.h.264-DOC is 720p WEB-DL
2015-09-17 11:37:09 INFO     SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Quality of Peep.Show.S07.720p.WEB-DL.AAC2.0.H.264-DOC is 720p WEB-DL
2015-09-17 11:37:09 INFO     SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Quality of Peep.Show.S06.720p.WEB-DL.AAC2.0.H.264-DOC is 720p WEB-DL
2015-09-17 11:37:09 INFO     SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Quality of Peep.Show.S04.SDTV.Xvid is Unknown
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Peep.Show.S04.SDTV.Xvid is a quality we know we don't want, rejecting it
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Picked Peep.Show.S07.720p.WEB-DL.AAC2.0.h.264-DOC as the best
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: The highest quality of any match is 720p WEB-DL
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: The quality of the season nzb is 720p WEB-DL
2015-09-17 11:37:09 INFO     SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Executed query: [SELECT episode FROM tv_episodes WHERE showid = 71656 AND season in  1,1,1,1,1,1]
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Episode list: [1, 2, 3, 4, 5, 6]
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Checking if found episode Peep Show S01E01 is wanted at quality 720p WEB-DL
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Any,Best = [ 720p HDTV, RawHD, 1080p HDTV, 720p WEB-DL, 1080p WEB-DL, 720p BluRay, 1080p BluRay ] [ RawHD, 1080p HDTV, 720p WEB-DL, 1080p WEB-DL, 720p BluRay, 1080p BluRay ] Found = [ 720p WEB-DL ]
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Existing episode status: 104 (Downloaded (SDTV))
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Quality is on wanted list, need to check if it's better than existing quality
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Episode already exists but the found episode quality is wanted more, getting found episode
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Checking if found episode Peep Show S01E02 is wanted at quality 720p WEB-DL
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Any,Best = [ 720p HDTV, RawHD, 1080p HDTV, 720p WEB-DL, 1080p WEB-DL, 720p BluRay, 1080p BluRay ] [ RawHD, 1080p HDTV, 720p WEB-DL, 1080p WEB-DL, 720p BluRay, 1080p BluRay ] Found = [ 720p WEB-DL ]
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Existing episode status: 104 (Downloaded (SDTV))
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Quality is on wanted list, need to check if it's better than existing quality
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Episode already exists but the found episode quality is wanted more, getting found episode
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Checking if found episode Peep Show S01E03 is wanted at quality 720p WEB-DL
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Any,Best = [ 720p HDTV, RawHD, 1080p HDTV, 720p WEB-DL, 1080p WEB-DL, 720p BluRay, 1080p BluRay ] [ RawHD, 1080p HDTV, 720p WEB-DL, 1080p WEB-DL, 720p BluRay, 1080p BluRay ] Found = [ 720p WEB-DL ]
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Existing episode status: 104 (Downloaded (SDTV))
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Quality is on wanted list, need to check if it's better than existing quality
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Episode already exists but the found episode quality is wanted more, getting found episode
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Checking if found episode Peep Show S01E04 is wanted at quality 720p WEB-DL
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Any,Best = [ 720p HDTV, RawHD, 1080p HDTV, 720p WEB-DL, 1080p WEB-DL, 720p BluRay, 1080p BluRay ] [ RawHD, 1080p HDTV, 720p WEB-DL, 1080p WEB-DL, 720p BluRay, 1080p BluRay ] Found = [ 720p WEB-DL ]
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Existing episode status: 104 (Downloaded (SDTV))
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Quality is on wanted list, need to check if it's better than existing quality
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Episode already exists but the found episode quality is wanted more, getting found episode
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Checking if found episode Peep Show S01E05 is wanted at quality 720p WEB-DL
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Any,Best = [ 720p HDTV, RawHD, 1080p HDTV, 720p WEB-DL, 1080p WEB-DL, 720p BluRay, 1080p BluRay ] [ RawHD, 1080p HDTV, 720p WEB-DL, 1080p WEB-DL, 720p BluRay, 1080p BluRay ] Found = [ 720p WEB-DL ]
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Existing episode status: 104 (Downloaded (SDTV))
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Quality is on wanted list, need to check if it's better than existing quality
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Episode already exists but the found episode quality is wanted more, getting found episode
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Checking if found episode Peep Show S01E06 is wanted at quality 720p WEB-DL
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Any,Best = [ 720p HDTV, RawHD, 1080p HDTV, 720p WEB-DL, 1080p WEB-DL, 720p BluRay, 1080p BluRay ] [ RawHD, 1080p HDTV, 720p WEB-DL, 1080p WEB-DL, 720p BluRay, 1080p BluRay ] Found = [ 720p WEB-DL ]
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Existing episode status: 104 (Downloaded (SDTV))
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Quality is on wanted list, need to check if it's better than existing quality
2015-09-17 11:37:09 DEBUG    SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Episode already exists but the found episode quality is wanted more, getting found episode
2015-09-17 11:37:09 INFO     SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Every ep in this season is needed, downloading the whole nzb Peep.Show.S07.720p.WEB-DL.AAC2.0.h.264-DOC
2015-09-17 11:37:09 INFO     SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Downloading Peep.Show.S07.720p.WEB-DL.AAC2.0.h.264-DOC from NZB.su
2015-09-17 11:37:09 INFO     SEARCHQUEUE-BACKLOG-71656 :: [NZB.su] :: Sending NZB to SABnzbd

**Comment by miigotu** [comment]
I agree that is a mess there, but I dont think the problem is the lines you noted. For one we shouldnt need to even query the db there, wantEpisode does that for us. Seems the code can match found episodes with wrong seasons for the check there.

**Comment by bobbysteel** [comment]
Is there any documentation on the data structures anywhere? I didn't see anything technical on the wiki and while doable, will take me a lot of time to get up to speed to start checking in code myself...

**Comment by miigotu** [comment]
Python documents itself =P

Seriously though, SR is in a constant state of rewrite, so any deep documentation would be nearly impossible to maintain.

**Comment by bobbysteel** [comment]
:) Fair enough. Trying to understand some of the flow here - at least with the first part of the bug above, is the intended behavior of pickBestResult to iterate over all possible season packs, or is it meant to be singular to pick the best season pack for any one season? I'm not sure whether that section should instead be in a loop for each particular season's pack (among possibly several seasons) in a result set.

**Comment by miigotu** [comment]
Honestly, I broke that part awhile back I think when fixing another issue.

pickBestResult is supposed to pick the best result out of a set of results, after filtering it based on criteria like desired quality for the show, ignored and required words, etc.

It originally was only used for episode results.

The problem here is that the list containing season results has results for other seasons in it than what was searched for (lazy matching from the search provider). This means that sometimes a season result that is not for the season you searched for ends up being picked instead of one that exists in your list that you did search for.

I think a quick hack around it is possible, but just puts a tiny band-aid on a gushing wound.

**Comment by bobbysteel** [comment]
Seems like the wantEpisode issue is that the way it's structured now, it needs to have a list of all episodes in the season which it then passes to wantEpisode to check if wanted specifically. Probably easiest to just create a wantSeason function I'd think that can return a bool value and iterate over each particular episode.

**Comment by miigotu** [comment]
I disagree, I think pickBestResult needs a seasons parameter, to filter out results that do not pertain to that particular result.

If it were wantEpisode that were changed, and the bestSeasonResult was for the wrong season, the correct season would never be parsed ^

**Comment by bobbysteel** [comment]
I'm not sure how to extract the individual episodes though to check with wantEpisode w/o a requisite DB query there though. Is there an episode list somewhere else already in an object I'm not familar with?

**Comment by miigotu** [comment]
the episodes the search was originated for are in a list of episode objects in the variable 'episodes', passed in as a parameter.

as for the season result (bestSeasonResult) the season it is for should be in bestSeasonResult.season

So if bestSeasonResult.season not in [x.season for x in episodes]: continue

**Comment by bobbysteel** [comment]
Hm. Shouldn't there be a loop there then? If you have results for multiple full seasons and eps are wanted from both seasons, I'd assume you'd want to have muliple passes over the bestSeasonResult output right?

**Comment by miigotu** [comment]
pickBestResult loops over the season results, and returns just one of the season results. If there are results for more than one season, that one result it picks may be the wrong result for the season we intended to search for.

**Comment by bobbysteel** [comment]
Maybe I'm misunderstanding the logic here. The episodes object passed in may have multiple seasons worth of episodes needed. When it comes down to looking for season results, what happens if only one  season is returned? Is there another outer loop somewhere? For ex, if all of season 3 and all of season 4 are needed, the code as I read it would only return the best result, leaving one of the other seasons unfound. Right? (Thx for the explanation here btw - just trying to get a mental map of the flow)

**Comment by miigotu** [comment]
the list of episodes passed in will not have more than one season. the results might.

**Comment by bobbysteel** [comment]
Ah that makes better sense then. Thanks
